tests/test_stamp.py

```diff
@@ -571,7 +571,51 @@ def new_toplevel_only(self, object_types):
             np.testing.assert_allclose(realized_X, predict, rtol=rtol)
 
 
+def test_lsst_silicon_builder_calculates_centroid() -> None:
+    """
+    LSST_SiliconBuilder.draw passes the correct list of photon_ops to
+    prof.drawImage.
+
+    This fails because no flux is actually put in the image, but I'm
+    not sure why
+    """
+    lsst_silicon = create_test_lsst_silicon(faint=False)
+    image = galsim.Image(ncol=256, nrow=256)
+    offset = galsim.PositionD(0,0)
+    config = create_test_config()
+    config['stamp']['centroid'] = {
+        'n_photons': 1000,
+    }
+
+    prof = galsim.Gaussian(sigma=2) * galsim.SED('vega.txt', 'nm', 'flambda')
+    method = 'phot'
+    logger = galsim.config.LoggerWrapper(None)
+
+    if galsim.__version_info__ < (2,5):
+        phot_type = 'galsim.ChromaticTransformation'
+    else:
+        # In GalSim 2.5+, this is now a SimpleChromaticTransformation
+        phot_type = 'galsim.SimpleChromaticTransformation'
+
+    with mock.patch(phot_type+'.drawImage', return_value=image) as mock_drawImage:
```

Comment:
Don't use mock.  Have this call the real drawImage function, and then it will produce an actual output image.  The above tests that use mock were testing something very specific about how parameters were being passed to drawImage, which mock lets you test.  I think you want to actually call the drawImage function and check that the centroid values get computed correctly.

Reply:
Yeah, I didn't know what that context did, but I found I had to  use it because otherwise I got this error:
```
imsim/stamp.py:778: in draw
    image = gal.drawImage(bp_for_drawImage,
../../miniforge3/envs/stack/lib/python3.11/site-packages/galsim/chromatic.py:2136: in drawImage
    image = ChromaticObject.drawImage(self, bandpass, image, integrator, **kwargs)
../../miniforge3/envs/stack/lib/python3.11/site-packages/galsim/chromatic.py:485: in drawImage
    image = prof0.drawImage(image=image, **kwargs)
../../miniforge3/envs/stack/lib/python3.11/site-packages/galsim/gsobject.py:1775: in drawImage
    added_photons, photons = prof.drawPhot(imview, gain, add_to_image,
../../miniforge3/envs/stack/lib/python3.11/site-packages/galsim/gsobject.py:2383: in drawPhot
    op.applyTo(photons, local_wcs, rng)
imsim/photon_ops.py:90: in applyTo
    v = self.photon_velocity(photon_array, local_wcs, rng)
imsim/photon_ops.py:189: in photon_velocity
    return self.rubin_diffraction.photon_velocity(
imsim/photon_ops.py:267: in photon_velocity
    v = photon_velocity(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

photon_array = galsim.PhotonArray(100, x=array([-4.95703380547105, -1.7774137125821585, 5.4679574148088355, 1.3853752380310527, -11.0...28, 567.8319745320754, 610.7805936794955, 588.7526624330898, 631.0916929335601, 659.
8606846139876, 662.0577956513312]))
xy_to_v = <imsim.photon_ops.XyToV object at 0x7165d0d3b350>, get_n = <bound method Medium.getN of ConstMedium(1.0)>

    def photon_velocity(photon_array, xy_to_v: "XyToV", get_n) -> np.ndarray:
        """Computes the velocity of a photon array."""
    
>       assert photon_array.hasAllocatedPupil()
E       AssertionError

imsim/photon_ops.py:123: AssertionError

```

Reply:
You need to either use an atmosphere photon_op or PupilAnnulusSampler.  The RubinDiffractionOptics photon_op requires that something else has already populated the pupil plane positions.  In normal imsim usage, it's the AtmosphericPSF.  But you can also do it manually with PupilAnnulusSampler (or PupilImageSampler works too, but you probably don't want to use that one.)

Reply:
I also had to add a TimeSampler

Reply:
Yeah, that sounds right.